File: hooks/node/src/schema-check.ts
// A validator for the subset of JSON Schema the event contract uses.
//
// Test utility, and the counterpart of hooks/python/tests/schema_check.py. It
// lives beside the sources because this package keeps its tests there too, and
// it is shared rather than copied so that the in-process check and the child
// process check are provably the same check.
//
// Reaching for ajv would mean a dependency, and the schema uses six keywords:
// type, required, properties, additionalProperties, enum, pattern, items,
// minimum and maximum. It is trusted only because the suite makes it prove
// itself: it has to accept the repository's valid example and reject the
// invalid one.

export type Schema = Record<string, unknown>;

/** Returns a list of error strings. Empty means the instance conforms. */
export function validate(schema: Schema, value: unknown, path = "$"): string[] {
  const errors: string[] = [];

  const enumeration = schema["enum"] as unknown[] | undefined;
  if (enumeration !== undefined && !enumeration.includes(value)) {
    errors.push(`${path}: ${String(value)} is not one of ${enumeration.join(", ")}`);
  }

  const type = schema["type"] as string | undefined;
  if (type === "object") {
    if (typeof value !== "object" || value === null || Array.isArray(value)) {
      return [`${path}: expected object`];
    }
    const record = value as Record<string, unknown>;
    const properties = (schema["properties"] ?? {}) as Record<string, Schema>;

    for (const key of (schema["required"] ?? []) as string[]) {
      if (!(key in record)) errors.push(`${path}.${key}: required`);
    }
    if (schema["additionalProperties"] === false) {
      for (const key of Object.keys(record)) {
        if (!(key in properties)) errors.push(`${path}.${key}: not allowed by the schema`);
      }
    }
    for (const [key, subSchema] of Object.entries(properties)) {
      if (key in record) errors.push(...validate(subSchema, record[key], `${path}.${key}`));
    }
    return errors;
  }

  if (type === "array") {
    if (!Array.isArray(value)) return [`${path}: expected array`];
    const items = schema["items"] as Schema | undefined;
    if (items !== undefined) {
      value.forEach((item, index) => errors.push(...validate(items, item, `${path}[${index}]`)));
    }
    return errors;
  }

  if (type === "string") {
    if (typeof value !== "string") return [`${path}: expected string`];
    const pattern = schema["pattern"] as string | undefined;
    if (pattern !== undefined && !new RegExp(pattern).test(value)) {
      errors.push(`${path}: ${value} does not match ${pattern}`);
    }
    return errors;
  }

  if (type === "integer") {
    if (!Number.isInteger(value)) return [`${path}: expected integer`];
    const minimum = schema["minimum"] as number | undefined;
    const maximum = schema["maximum"] as number | undefined;
    if (minimum !== undefined && (value as number) < minimum) errors.push(`${path}: below minimum`);
    if (maximum !== undefined && (value as number) > maximum) errors.push(`${path}: above maximum`);
  }

  return errors;
}

File: hooks/python/tests/schema_check.py
"""A small JSON Schema checker, limited to the keywords the event schema uses.

The hook may not depend on anything outside the standard library, and neither
may its tests. This covers `type`, `required`, `properties`,
`additionalProperties: false`, `enum`, `pattern`, `items`, `minimum` and
`maximum`, which is every keyword in egress-event.schema.json.

It is trusted only because the suite makes it prove itself: it has to accept the
repository's valid example and reject the invalid one for the documented reason
(schemas/examples/invalid-expectations.json).
"""

import re

_TYPES = {
    "object": dict,
    "array": list,
    "string": str,
    "boolean": bool,
    "null": type(None),
}


def _type_matches(value, name):
    if name == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if name == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    expected = _TYPES.get(name)
    if expected is None:
        return True
    if expected is not bool and isinstance(value, bool):
        return False
    return isinstance(value, expected)


def validate(instance, schema, path=""):
    """Return a list of error strings. Empty means the instance conforms."""
    errors = []
    declared = schema.get("type")
    if declared is not None:
        names = declared if isinstance(declared, list) else [declared]
        if not any(_type_matches(instance, name) for name in names):
            errors.append("{0}: expected type {1}".format(path or "/", declared))
            return errors

    if "enum" in schema and instance not in schema["enum"]:
        errors.append("{0}: {1!r} is not one of {2}".format(
            path or "/", instance, schema["enum"]))

    if "pattern" in schema and isinstance(instance, str):
        if re.search(schema["pattern"], instance) is None:
            errors.append("{0}: {1!r} does not match {2}".format(
                path or "/", instance, schema["pattern"]))

    if isinstance(instance, (int, float)) and not isinstance(instance, bool):
        if "minimum" in schema and instance < schema["minimum"]:
            errors.append("{0}: below minimum".format(path or "/"))
        if "maximum" in schema and instance > schema["maximum"]:
            errors.append("{0}: above maximum".format(path or "/"))

    if isinstance(instance, dict):
        errors.extend(_validate_object(instance, schema, path))
    elif isinstance(instance, list) and "items" in schema:
        for index, item in enumerate(instance):
            errors.extend(
                validate(item, schema["items"], "{0}/{1}".format(path, index))
            )
    return errors


def _validate_object(instance, schema, path):
    errors = []
    properties = schema.get("properties", {})
    for name in schema.get("required", []):
        if name not in instance:
            errors.append("{0}/{1}: required property missing".format(path, name))
    for name, value in instance.items():
        child = "{0}/{1}".format(path, name)
        if name in properties:
            errors.extend(validate(value, properties[name], child))
        elif schema.get("additionalProperties") is False:
            errors.append("{0}: property not allowed".format(child))
    return errors
